### PR TITLE
implemented Interchip 

### PR DESCRIPTION
**Pull request by SnackkOverflowError**
Interchip autopilot is called every 20 ms in a freertos thread (not in this PR, it will be added in another PR). 

Interchip Safety is called once before the main loop, and each interchip callback calls the HAL_SPI_TransmitReceive function, which keeps safety listening and sending data.

Safety has functions to get PWM, and get/setsafety level.
Autopilot has functions to set PWM and get safety level.

### CODE REVIEW

File: interchip-Autopilot/Core/Inc/Interchip.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
```
Comment (Fberbari) [inline interchip-Autopilot/Core/Inc/Interchip.hpp]: You meant to get rid of all these files, right ?

File: Safety/boardfiles/Safety_ZeropilotRev1.ioc [hunk 2]
```diff
@@ -24,22 +24,22 @@ Mcu.Name=STM32F030RCTx
...
-Mcu.Pin10=PC8
-Mcu.Pin11=PC9
-Mcu.Pin12=PA8
-Mcu.Pin13=PA9
-Mcu.Pin14=PA10
-Mcu.Pin15=PA11
-Mcu.Pin16=PA13
-Mcu.Pin17=PA14
-Mcu.Pin18=PA15
-Mcu.Pin19=PC10
```
Comment (Fberbari) [inline Safety/boardfiles/Safety_ZeropilotRev1.ioc:36]: I'm perplexed.... Why are pins being renamed ?

File: Safety/Src/interchip_S.cpp [hunk 3]
```diff
@@ -1,93 +1,57 @@
...
-    if (transmit_status != HAL_OK) {
-        return 1;
-    }
-    else
-    {
-        return 0;
-    }
+// Starts interchip interrupt. Call this once, not in a loop, each interchip callback
+// sets up another interrupt.
+void sendReceiveData() {
```
Comment (Fberbari) [inline Safety/Src/interchip_S.cpp]: THis should be done in some sort of init function, The name is misleading.

File: Safety/Src/interchip_S.cpp [hunk 4]
```diff
@@ -1,93 +1,57 @@
...
-
-static Interchip_StoA_Packet* dataTX;
-static Interchip_AtoS_Packet* dataRX;
-static uint16_t errorCount=0;
-//
-
-uint16_t getErrorCount(){
-  return errorCount;
+// Get the pwm signal that has been recieved.
+int16_t getPWM(int8_t index) {
```
Comment (Fberbari) [inline Safety/Src/interchip_S.cpp]: Can you actually just return the whole struct 1 shot ? Makes things a bit easier, also easier in terms of the isDataNew stuff, which this function may as well return.

File: Autopilot/boardfiles/Autopilot_ZeropilotRev1.ioc [hunk 5]
```diff
@@ -71,55 +72,59 @@ Mcu.Pin15=PA0/WKUP
...
-Mcu.Pin21=PF14
-Mcu.Pin22=PF15
-Mcu.Pin23=PG0
-Mcu.Pin24=PG1
-Mcu.Pin25=PE7
-Mcu.Pin26=PE13
-Mcu.Pin27=PB10
-Mcu.Pin28=PB11
-Mcu.Pin29=PB12
+Mcu.Pin20=PA5
```
Comment (Fberbari) [inline Autopilot/boardfiles/Autopilot_ZeropilotRev1.ioc:77]: really unsure of what's going on here.... Kinda making me nervous.

File: Autopilot/Src/imu.cpp [hunk 6]
```diff
@@ -1,4 +1,5 @@
 #include "imu.hpp"
+#include "Interchip_A.h"
```
Comment (Fberbari) [inline Autopilot/Src/imu.cpp]: This is a brutal solution. Just make a file called "Callbacks.cpp" and stick in AutoPilot/src. I'm making a note in the asana that this should be done, but no need to address now, cuz we simply don't have the luxury, but the callbacks file shoiuld just call some appropraite callback function inside each module.

**Review by Fberbari — COMMENTED** [review]
There's a lot of cosmetic stuff... But not now, after comp. This looks good, but do you know anything about why cubeMX is renaming pins ?? That just seems really weird. Also gotta get rid of all the interchip-autopilot crap

File: interchip-Autopilot/Core/Inc/Interchip.hpp (hunk 1 above)
Comment (SnackkOverflowError) [reply interchip-Autopilot/Core/Inc/Interchip.hpp]: did not realize I commited those, whoops

File: Autopilot/boardfiles/Autopilot_ZeropilotRev1.ioc (hunk 5 above)
Comment (SnackkOverflowError) [reply Autopilot/boardfiles/Autopilot_ZeropilotRev1.ioc:77]: Me too actually, Ill investigate this further

File: Autopilot/Src/imu.cpp (hunk 6 above)
Comment (SnackkOverflowError) [reply Autopilot/Src/imu.cpp]: Im not sure if my fix is more or less hacky than what is currently there, but now the callback code is all in Callbacks.cpp

File: Safety/Src/interchip_S.cpp (hunk 4 above)
Comment (SnackkOverflowError) [reply Safety/Src/interchip_S.cpp]: oh yeah I hadnt considered the isDataNew stuff, will do. Should I be returning the entire struct or just the PWM array?

Comment (SnackkOverflowError) [reply Safety/Src/interchip_S.cpp]: it now returns the array

File: Autopilot/Src/Callbacks.cpp [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#include "Interchip_A.h"
+#include "imu.hpp"
+#include "spi.h"
+
+void HAL_SPI_TxRxCpltCallback(SPI_HandleTypeDef *hspi)
+{
+    if(hspi == &hspi1) {
```
Comment (Fberbari) [inline Autopilot/Src/Callbacks.cpp:7]: This is the right idea. Not hacky, it's good.

For reference, the really proper solution is to have each module pass the callback module a function pointer to call. That way, the callback module doesn't need to #include anyone and also that callback prototypes don't need to be exposed to the higher level. but that's of very low value and this solution is almost just as good.

File: Autopilot/Inc/Interchip_A.h [hunk 8]
```diff
@@ -10,12 +10,14 @@
...
 extern "C" {
 #endif
+ 
 
+void Interchip_Init();
 void Interchip_Run();
-int16_t *Interchip_GetPWM(void);
-void Interchip_SetPWM(int16_t *data);
+void Interchip_SetPWM(int index,int data);
 uint16_t Interchip_GetAutonomousLevel(void);
```
Comment (Fberbari) [inline Autopilot/Inc/Interchip_A.h:21]: Each of these will need a comment to explain it. As we do with all our function prototypes.

File: Autopilot/Inc/imu.hpp [hunk 9]
```diff
@@ -84,4 +84,6 @@ class BMX160: public IMU{
 
 };
 
+void ImuTxRxCallback();
```
Comment (Fberbari) [inline Autopilot/Inc/imu.hpp:92]: Add this above this prototype pls:

        /**
         * Callback for exclusive use by the SPI ISR.
         * Users of this module should never call this function.
         * Any manipulation of this function by the higher level voids all guarantees made by this module.
         * */

**Review by Fberbari — COMMENTED** [review]
Sweet, couple more small things. the only hurdle is that IOC garbage, but I'm gonna take a look at that as well.

File: Common/Inc/Interchip.h [hunk 10]
```diff
@@ -1,17 +1,12 @@
...
-} Interchip_StoA_Packet;    //Safety to Autopilot packet
 
 typedef struct {
+	uint16_t safetyLevel;
 	int16_t PWM[12];
-	uint16_t autonomous_level;
-} Interchip_AtoS_Packet;    //Autopilot to Safety packet
+} Interchip_Packet;  
 
 #endif
```
Comment (Pop0097) [inline Common/Inc/Interchip.h:11]: newline . See if there are any other places where the "do not enter" sign appears (we really need to automate this lol)

File: Safety/Src/interchip_S.cpp [hunk 11]
```diff
@@ -1,93 +1,57 @@
...
+	txData.safetyLevel = 13; 
 }
 
-#endif
-
+// spi callback that starts up another interrupt.
+void HAL_SPI_TxRxCpltCallback(SPI_HandleTypeDef* hspi) {
+	// Start Listening after the previous exchange.
+	HAL_SPI_TransmitReceive_IT(&hspi1,(uint8_t *)&txData,(uint8_t *)&rxData, sizeof(Interchip_Packet));
+	dataNew = true;
```
Comment (Pop0097) [inline Safety/Src/interchip_S.cpp:56]: smart move

**Review by Pop0097 — COMMENTED** [review]
Not much for me to say beyond what Anthony mentioned.

**Review by Fberbari — APPROVED** [review]

**Review by Pop0097 — APPROVED** [review]
This is tested to work and the code looks good to me. Let's integrate it and see how it flies!

**Pull request merged**